xain/grpc/participant.py

```diff
@@ -65,34 +118,240 @@ def end_training(
+#         xy_train,
+#         xy_val,
+#         num_classes=10,
+#         batch_size=FLAGS.batch_size,
+#     )
+
+
+def training_round(channel, participant):
+    """Initiates training round exchange with Coordinator.
+
+    Begins with `start_training`. Then performs local training computation using
+    `participant`. Finally, completes with `end_training`.
+
+    Args:
+        channel: gRPC channel to Coordinator.
+        participant (obj:`Participant`): Local Participant.
+    """
+    theta, epochs, base = start_training(channel)
+    # training:
+    theta_n, his, _dict = participant.train_round(theta, epochs, base)
+    # NOTE _dict is the opt_config - ignore for now
+    met = participant.metrics()
+    end_training(channel, theta_n, his, met)
+
+
+# def main(_argv):
+#     print(f"model_name: {FLAGS.model_name}")
+#     print(f"dataset_name: {FLAGS.dataset_name}")
+#     print(f"batch_size: {FLAGS.batch_size}")
+#     print(f"partition_iden: {FLAGS.partition_iden}")
+#     go(init_participant())
+
+
+# if __name__ == "__main__":
+#     flags.mark_flag_as_required("model_name")
+#     flags.mark_flag_as_required("dataset_name")
+#     flags.mark_flag_as_required("batch_size")
+#     flags.mark_flag_as_required("partition_iden")
+#     app.run(main)
+
+
+class StateRecord:
+    """Thread-safe record of Participant state and round number.
+    """
+
+    def __init__(self, state=ParState.WAITING_FOR_SELECTION, round=0):
+        self.cv = threading.Condition()
+        self.round = round
+        self.state = state
+
+    def lookup(self):
+        """Looks up the state and round number.
+
+        Returns:
+            :obj:`Tuple[ParState, int]`: State and round number
+        """
+        with self.cv:
+            return self.state, self.round
+
+    def update(self, state):
+        """Updates state.
+
+        Args:
+            state (:obj:`ParState`): State to update to.
+        """
+        with self.cv:
+            self.state = state
+            self.cv.notify()
+
+    def wait_until_selected_or_done(self):
+        """Waits until Participant is in the state of having been selected for training
+        (or is completely done).
+
+        Returns:
+            :obj:`ParState`: New state Participant is in.
+        """
+        with self.cv:
+            self.cv.wait_for(lambda: self.state in {ParState.TRAINING, ParState.DONE})
+            # which one was it?
+            return self.state
+
+    def wait_until_next_round(self):
+        """Waits until Participant is in a state indicating the start of the next round
+        of training.
+
+        Returns:
+            :obj:`ParState`: New state Participant is in.
+        """
+        with self.cv:
+            self.cv.wait_for(
+                lambda: self.state
+                in {ParState.TRAINING, ParState.WAITING_FOR_SELECTION, ParState.DONE}
+            )
+            # which one was it?
+            return self.state
+
+
+def transit(st, beat_reply):
```

Comment:
For future reference I think that we could use Jan's suggestion of removing the POST_TRAINING state and maybe reuse only the proto defined states `READY, TRAINING and FINISHED`

Reply:
Yes I have a proposal for reducing the participant states down to two making it a bit simpler conceptually